lib/milestone_date_change.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+class MilestoneDateChange
+  class DateCannotBeChangedError < RuntimeError; end
+
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+  include ActiveModel::Validations
+
+  attribute :schedule_identifier, :string
+  attribute :start_year, :integer
+  attribute :milestone_number, :integer
+  attribute :new_start_date, :date
+  attribute :new_milestone_date, :date
+
+  validates :new_milestone_date, presence: { message: I18n.t(:milestone_date_required) }, if: -> { new_start_date.blank? }
+  validates :new_start_date, presence: { message: I18n.t(:milestone_date_required) }, if: -> { new_milestone_date.blank? }
+  validates :milestone, presence: { message: I18n.t(:milestone_not_matched) }
+  validate :validate_milestone_dates_present
+  validate :validate_milestone_date_change
+
+  def change_date!
+    raise DateCannotBeChangedError, errors.map(&:message) if invalid?
+
+    updates = { start_date: new_start_date, milestone_date: new_milestone_date }.compact
+    milestone.update!(updates)
+  end
+
+  def milestone
+    return nil unless schedule
+
+    @milestone ||= schedule.milestones.find do |m|
+      m.name.start_with?("Output #{milestone_number} - ")
+    end
+  end
+
+  def milestone_declarations
+    return [] unless milestone
+
+    ParticipantDeclaration::ECF
+      .includes(:participant_profile)
+      .where(participant_profile: { schedule_id: schedule.id })
+      .where(declaration_date: existing_date_range)
```

Comment:
could you append `.where.not(state: ["voided", "awaiting_clawback", "clawed_back"])` to the query to ensure we don't get any voided declarations which we won't want to include in our validations?